### PR TITLE
7979 - Publish Cloud event when NR change state

### PR DESCRIPTION
**Pull request by andrepestana-aot**
*Issue #, if available: #7979*

*Description of changes:*

- Added CloudEventMessageService; 
- Added new configurations; 
- Added Request model event listener.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namex license (Apache 2.0).

### CODE REVIEW

File: api/namex/services/cloud_event_messages/cloud_event_messages.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                                                utcfromtimestamp(time.time()).
+                                                replace(tzinfo=timezone.utc).
+                                                isoformat(),
+                                                identifier=nr_num,
+                                                json_data_body={
+                                                    'request': {
+                                                        'nrNum': nr_num,
+                                                        'oldNrNum': old_nr_num,
+                                                        'newState': state_cd,
+                                                        'previousState': old_state_cd
```
Comment (andrepestana-aot) [inline api/namex/services/cloud_event_messages/cloud_event_messages.py:53]: Included new and old values as @sumesh-aot pointed out that the update from NRL to NR may be necessary:
https://github.com/bcgov/entity/pull/8003/files#r660844340

Comment (sumesh-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:53]: Do you store the account-id anywhere in names? To auto-affiliate the NR to the account, we will need to know the account to affiliate.

Comment (andrepestana-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:53]: Do you mean the account-id from the user who created the NR?

Comment (sumesh-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:53]: Yes, if it's an authenticated user.

Comment (sumesh-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:53]: Another option for us is to get the account information from Pay records for the NR.

File: api/namex/services/cloud_event_messages/cloud_event_messages.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                                                source=f'/requests/{nr_num}',
+                                                time=datetime.
+                                                utcfromtimestamp(time.time()).
+                                                replace(tzinfo=timezone.utc).
+                                                isoformat(),
+                                                identifier=nr_num,
+                                                json_data_body={
+                                                    'request': {
+                                                        'nrNum': nr_num,
+                                                        'oldNrNum': old_nr_num,
```
Comment (thorwolpert) [inline api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: Why is there an old NR num here?
The NRL is and only the NR is there when an invoice is created.
I'm not sure of the flow, but are we going to affiliate NRs that haven't been paid for?

**Review by thorwolpert — CHANGES_REQUESTED** [review]
Might be worth copy/pasting in https://github.com/bcgov/lear/blob/main/legal-api/src/legal_api/services/queue.py

**Comment by thorwolpert** [comment]
Looks like @vysakh-menon-aot already brought it in.
You'll need to rebase to pickup https://github.com/bcgov/lear/blob/main/legal-api/src/legal_api/services/queue.py

File: api/namex/services/cloud_event_messages/cloud_event_messages.py (hunk 2 above)
Comment (andrepestana-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: I can remove that. I'm just trying to satisfy what Sumesh pointed out: https://github.com/bcgov/namex/pull/1138#discussion_r671594620

Comment (andrepestana-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: Please @sumesh-aot confirm we don't need the temporary NRL.

Comment (sumesh-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: @andrepestana-aot @thorwolpert Not sure if NRL is being used now. 
Old and new identifier is needed if the name request uses NRL while creating an invoice, then only we can update the identifier to new identifier to properly show in receipt.

Comment (thorwolpert) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: An NR doesnt have both an NRL an a NR# so there's no way to map between the 2
It'll depend on if you affiliate an NR that hasn't been paid for yet.
we remove those after 24hrs, so it probably isn't worth it?

Comment (sumesh-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: Affiliation will be based on the NR status change. So I don't think affiliation would happen before payment is received. But the need for old and new was for displaying in receipts and other payment records. 
- If the invoice was created with the NRL identifier, all the pay records would be having NRL. And if we need to update the NRL to NR, the event will need this info (similar to lear where the event contains temp registration number and permanent number).
- Also if we are planning to get the account information from pay records (one of my comment on this PR above) - we won't be able to get it if the records in pay stays NRL.

The second can be solved if the Names stores account-Id and send it in the event.

Comment (thorwolpert) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: I asked at standup, currently we wont be affiliating NRs that have been paid for, so we don't need to publish events for NRLs

Comment (thorwolpert) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: @sumesh-aot we're not there yet. it'll take more changes to support both NRL/NR# at the same time, as well as change the pay flow to be more like businesses (create the receipt with the temp reg, then change if paid/completed).

Comment (sumesh-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: Yes, that aligns with my understanding. So do we get account-id in event (if NRL is being used for invoice creation) ?

Comment (sumesh-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: aahh ok. Strike my above comment, it was a race condition :)

Comment (thorwolpert) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:51]: Looks correct to me.
i don't think it should publish events for states prior to DRAFT (which is the state you enter once paid).

File: api/namex/services/cloud_event_messages/cloud_event_messages.py (hunk 1 above)
Comment (andrepestana-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:53]: I see there's a user_id column for the `requests` and `events` tables but they are not populating with the user_id from the user who is creating the NR in namex database.

Comment (thorwolpert) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py:53]: it is if the user is logged in.

Comment (andrepestana-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py]: Yes, it is possible to get the logged user from the request but namex-api is not saving the logged user. It saves name request account service userId in `requests` and `events` tables.

Comment (thorwolpert) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py]: ex. non-authenticated user submitted NR
```
   nr_num   |    state_cd     |           username
------------+-----------------+------------------------------
 NR 3640335 | DRAFT           | name_request_service_account
 NR 3640335 | PENDING_PAYMENT | name_request_service_account
 NR 3640335 | PENDING_PAYMENT | name_request_service_account
 NR 3640335 | DRAFT           | name_request_service_account
 NR 3640335 | INPROGRESS      | idir\kyee
 NR 3640335 | INPROGRESS      | idir\kyee
 NR 3640335 | INPROGRESS      | idir\kyee
 NR 3640335 | APPROVED        | idir\kyee
 NR 3640335 | APPROVED        | nro_service_account
 ```
 Looks like non staff are not captured via namerequest ... wtf

Comment (thorwolpert) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py]: in any case, first emitted message is in the DRAFT state, which is after payment.
so currently, PAY never has any NRLs, so there is no oldNrNum that is used.

Comment (andrepestana-aot) [reply api/namex/services/cloud_event_messages/cloud_event_messages.py]: Yes, I fixed that. Only NRLs changes are not sending messages.

**Comment by andrepestana-aot** [comment]
I'm creating a new PR because of the wrong merge to this branch

**Pull request closed without merging**